### PR TITLE
Allow filename and iPXE configuration in dhcp host entries

### PR DESCRIPTION
**Pull request by tom-krieger**
#### Pull Request (PR) description

DHCP allows to use a filename and a condition for iPXE within the DHCP host definition. Currently the module does not support this. Therefore I added three new optional parameters in the dhcp::host class and updated the host template to write these parameters into the configuration file. If the new parameters are not used there are no changes written to already existing dhcp configuration.

### CODE REVIEW

**Comment by kenyon** [comment]
Looks good, wonder if you could also a test for this in https://github.com/voxpupuli/puppet-dhcp/blob/master/spec/classes/dhcp_spec.rb ?

**Comment by tom-krieger** [comment]
> Looks good, wonder if you could also a test for this in https://github.com/voxpupuli/puppet-dhcp/blob/master/spec/classes/dhcp_spec.rb ?

Ok np., will do

**Comment by tom-krieger** [comment]
Added tests to [host_spec.rb](https://github.com/voxpupuli/puppet-dhcp/blob/master/spec/defines/host_spec.rb) as the changes are affecting only a dhcp host.

**Review by kenyon — APPROVED** [review]

File: manifests/host.pp [hunk 1]
```diff
@@ -9,6 +9,9 @@
   Boolean $ignored                      = false,
   Optional[Integer] $default_lease_time = undef,
   Optional[Integer] $max_lease_time     = undef,
+  Optional[String[1]] $ipxe_filename    = undef,
+  Optional[String[1]] $ipxe_bootstrap   = undef,
+  Optional[String[1]] $filename         = undef,
```
Comment (ekohl) [inline manifests/host.pp:14]: Having parameter documentation for every newly added parameter would be nice.

File: templates/dhcpd.host.erb [hunk 2]
```diff
@@ -42,4 +45,12 @@ host <%= @host %> {
 <% end -%>
   }
 <% end -%>
+<% if @ipxe_filename and @ipxe_bootstrap -%>
+  if exists <%= @ipxe_filename %> {
```
[the remark this answers is not in the document]
Comment (tom-krieger) [reply templates/dhcpd.host.erb:49]: Nope the code is correct. 

`if exists <%= @ipxe_filename %> {` is true when an iPXE client contacts the DHCP server, and false when a regular UEFI PXE boot ROM contacts the DHCP server. That's the reason for how the 'if' is structured. I do not want to hard code the names as there can be various ipxe clients.

File: manifests/host.pp (hunk 1 above)
Comment (tom-krieger) [reply manifests/host.pp:14]: Where shall I put down the documentation?

Comment (ekohl) [reply manifests/host.pp:14]: Ideally as puppet-strings inside this class. So
```puppet
# @summary Define a DHCP host
#
# @param ipxe_filename ...
# @param ipxe_bootstrap ...
define dhcp::host (
```

I won't make you document the existing parameters (though that would be awesome), but if we can at least document new parameters we can eventually get to a point where we can generate a useful reference documentation.

Comment (tom-krieger) [reply manifests/host.pp:57]: I just pushed the parameter documentation.

File: manifests/host.pp [hunk 3]
```diff
@@ -1,5 +1,48 @@
 # == Define: dhcp::host
 #
+# @param ip
+#    The ip address of the DHCP host or a DNS resolvable name.
```
[the remark this answers is not in the document]
Comment (tom-krieger) [reply manifests/host.pp:4]: Sorry my bad. DHCP does accept an dns resolvable name but the module you're right. I overlooked the 'Stdlib::IP::Address'. I fix the param.

File: manifests/host.pp [hunk 4]
```diff
@@ -1,5 +1,48 @@
 # == Define: dhcp::host
```
Comment (ekohl) [inline manifests/host.pp:1]: This is old rdoc style documentation. Would you mind changing this while you're at it?
```suggestion
# @summary Define a DHCP host
```

**Review by ekohl — APPROVED** [review]
No objection to merging.

File: templates/dhcpd.host.erb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ekohl) [reply templates/dhcpd.host.erb:49]: I must admit I'm not too familiar with iPXE but it feels odd to me to check if the filename exists and then serve the bootstrap. If it doesn't exist, then serve the filename.

[the remark this answers is not in the document]
Comment (tom-krieger) [reply templates/dhcpd.host.erb:49]: It's not the filename that is checked. The check is about the pxe client. It's unfortunately the same as the filename. I currently have no better idea on how to write that code in a better way or use better names for the template variables.

**Pull request merged**